gateway/src/main/java/io/camunda/zeebe/gateway/ResponseMapper.java

```diff
@@ -276,22 +278,53 @@ public static SetVariablesResponse toSetVariablesResponse(
     return SetVariablesResponse.newBuilder().setKey(key).build();
   }
 
-  public static ActivateJobsResponse toActivateJobsResponse(
-      final long key, final JobBatchRecord brokerResponse) {
+  /**
+   * While converting the broker response to the gRPC response, the response size is checked. If the
+   * response size exceeds the maximum response size, exceeding jobs are added to the list of
+   * exceeding jobs to be reactivated.
+   *
+   * <p>This is because the jobs returned from the broker is in MessagePack format and while
+   * converting them to gRPC response the size of the response may increase (e.g. we do JSON and
+   * String conversions see: {@link #toActivatedJob(long, JobRecord)}). That will cause the response
+   * size to exceed the maximum response size allowed by the gateway and the gateway will log a
+   * Stream Error indicating that streaming out the activated jobs failed.
+   *
+   * <p>If we do not respect the actual max response size, Zeebe Java Client rejects the response
+   * containing the activated jobs and the client cancels the channel/stream/connection as well.
+   * Leaving failed jobs non-activatable until their configured timeout.
+   *
+   * @param key the key of the request
+   * @param brokerResponse the broker response
+   * @param maxResponseSize the maximum size of the response
+   * @return a pair of the response and a list of jobs that could not be included in the response
+   *     because the response size exceeded the maximum response size
+   */
+  public static JobActivationResult toActivateJobsResponse(
+      final long key, final JobBatchRecord brokerResponse, final long maxResponseSize) {
     final ActivateJobsResponse.Builder responseBuilder = ActivateJobsResponse.newBuilder();
 
     final Iterator<LongValue> jobKeys = brokerResponse.jobKeys().iterator();
     final Iterator<JobRecord> jobs = brokerResponse.jobs().iterator();
 
+    long currentResponseSize = 0L;
+    final List<ActivatedJob> sizeExceedingJobs = new ArrayList<>();
+
     while (jobKeys.hasNext() && jobs.hasNext()) {
       final LongValue jobKey = jobKeys.next();
       final JobRecord job = jobs.next();
       final ActivatedJob activatedJob = toActivatedJob(jobKey.getValue(), job);
 
-      responseBuilder.addJobs(activatedJob);
+      // +2 is for the metadata size that is added to ActivateJobsResponse when a new ActivatedJob
+      // is added to it
+      currentResponseSize += activatedJob.getSerializedSize() + 2;
```

Comment:
🤔 This looks hard to maintain. If we ever add something to `ActivatedJob` or `ActivateJobsResponse` then this is no longer correct.

🔧 How about adding a post-condition check at the end of this method?
```java
if (currentResponseSize != response.getSerializedSize()) {
 // we've calculated the current response size incorrectly
}
```

Reply:
@korthout If that is the case, how are we going to handle it? Are we going to put all jobs to `sizeExceedingJobs`? Or take the last job from the responseBuilder jobs list and put it into `sizeExceedingJobs`?

Reply:
When that happens, it means we have a bug. We can:
- log an error so users can report the bug
- throw an exception and reject the command
- or somehow add a test case that fails when the size is calculated wrongly (seems difficult)

I don't see a need to complicate the job selection logic for this. If we calculate the size wrongly **and** the response exceeds the size limit, then this will result in an error again.

My personal preference is simply logging an error, hoping that a user reports it.

Reply:
@korthout The if check suggestion you provided will not do the work. Because current response size will always be different than the `response.getSerializedSize()` (e.g. serialized size will be +2).

Reply:
I would either keep `+2` or not add it at all. Even if a new field is added to the ActivateJobsResponse and we faced with exceeding size failures, it will be easier to add a fix for that. WDYT? @korthout

Reply:
>The if check suggestion you provided will not do the work. Because current response size will be different than the `response.getSerializedSize()` even if we add jobs to exceeding list.

I don't understand what you mean. 

❓ In my mind, the goal is to not exceed the max message size with the response because that may break things. We added a way to calculate the response size while adding jobs to the response. That calculation is 0 bytes for the response + for each job: 'size of the job + 2 bytes for metadata'. Just before returning the response, we can also request the actual size of the response with `response.getSerializedSize()` which should be equivalent to the manually calculated size. Or is that not the case? Am I missing something?

Reply:
Ohh :

Reply:
Ohh 😯. You mean keeping the `+2` and if we forget to update the value on field updates, we log the error. Then, we expected to see these errors in log and fix the issue. Is that correct? @korthout 

I thought it is hard to maintain +2, so we should remove it.

Reply:
> You mean keeping the +2 and if we forget to update the value on field updates, we log the error. Then, we expected to see these errors in log and fix the issue. Is that correct?

Yes, that's correct. I would keep that logic while it's correct, but add a check on top that tells us that the logic is correct. If things ever change, we might not have a failing test for this, but at least we have an ERROR log message that users could report.